### ISSUE TITLE
Pipeline/build prioritization

### ISSUE THREAD

**Issue by nikyoudale** [buildkite/feedback#147]
It would be great to be able to assign all build jobs from a particular pipeline a higher priority than jobs from another pipeline.

Example use case:
A 2 pipeline configuration: one pipeline builds commits on the master branch and is triggered automatically via github events, a second pipeline is used to build and deploy the app and is triggered manually (relatively infrequently). Jobs from both pipelines are runnable on all agents in a shared pool.

We have a setup like this and find that when it comes time to deploy, the jobs in the deploy pipeline often have to wait a very long time for an available agent in the pool (all busy most of the time building things in  'build master" pipeline). Even when busy agents finish their current job, they will often choose another job from the "build master" pipeline, rather than "deploy". This can create long lag times for a manual job we want to run asap. What we want to be able to do is configure any scheduled jobs from the deploy pipeline to always run before any scheduled jobs from the "build master" pipeline. Alternatively, a way to assign a priority to a build (e.g. an option on the create build API), such that jobs from that build are assigned ahead of other scheduled jobs with a lower, or absent, priority.

**Comment by toolmantim** [comment]
Thanks for the feedback!

Can I ask if you’ve looked at using build queues to achieve this? If you want to keep agents clear to do timely/urgent work (such as deploys) normally you'd use queues to achieve it: https://buildkite.com/docs/agent/queues

**Comment by nikyoudale** [comment]
Build queues would definitely work. The downside is that for this setup, as I understand it, we'd need to have at least one agent running tagged as an agent that runs *only* jobs from the deploy queue. If deploy jobs run only once per day, then having that agent sitting there doing nothing for the rest of the day is a bit of a waste.

**Comment by mikezucc** [comment]
Plus one for this, was there a resolution for this?

**Comment by Mattmlm** [comment]
Is there anyway to implement this with what Buildkite already provides?

**Comment by lvangool** [comment]
Adding another vote for this feature. Building the "X" branch pipeline queue as a priority over every other pipeline queue would be amazing! 
Running multiple agents has some concurrency issues in this specific case so we can't simply run a dedicated "x" branch agent.

**Comment by ccarpita** [comment]
Another vote for this feature! Build queues are insufficient on their own, because they essentially partition scarce resources. We would ideally allow our postmerge jobs to run on all agents at a lower priority, and have Pull Request builds jump the queue and run ASAP, since developers are waiting on those. Even if we allocate 50% of agents to a "postmerge" queue, it still reduces our capacity by 50% (which is bad for developers), and we'd want PR builds to jump the queue on those separate agents **in addition to** the default queue.

**Comment by staticfloat** [comment]
It would be really useful to be able to prioritize queues; In my use-case, an agent generally should belong to a set of "primary" queues, where it pulls work items from the primary queues in a round-robin fashion, and only if there are no work items in the set of primary queues does it pull a work item from one of the secondary queues.  This allows excess work from other queues to be run on workers that don't have any other pieces of work, while still allowing prioritization of important work, as specified by queues.

I'm sure there are many more complex prioritization schemes, but separating sets of queues as "primary"/"secondary" would fulfill all fo my use cases.

**Comment by joscha** [comment]
see https://buildkite.com/docs/pipelines/managing-priorities
